src/plugin/src/main/java/com/defenseunicorns/uds/keycloak/plugin/authentication/authenticators/client/ClientIdAndKubernetesSecretAuthenticator.java

```diff
@@ -0,0 +1,235 @@
+        secretMountPath = System.getenv(DEFAULT_ENVIRONMENT_VARIABLE_CLIENT_MOUNT_PATH);
+        if (secretMountPath == null) {
+            secretMountPath = DEFAULT_MOUNT_PATH;
+        }
+    }
+
+    @Override
+    public void authenticateClient(ClientAuthenticationFlowContext context) {
+        String client_id = null;
+        String clientSecret = null;
+
+        String authorizationHeader = context.getHttpRequest().getHttpHeaders().getRequestHeaders().getFirst(HttpHeaders.AUTHORIZATION);
+
+        MediaType mediaType = context.getHttpRequest().getHttpHeaders().getMediaType();
+        boolean hasFormData = mediaType != null && mediaType.isCompatible(MediaType.APPLICATION_FORM_URLENCODED_TYPE);
+
+        MultivaluedMap<String, String> formData = hasFormData ? context.getHttpRequest().getDecodedFormParameters() : null;
+
+        if (authorizationHeader != null) {
+            String[] usernameSecret = BasicAuthHelper.RFC6749.parseHeader(authorizationHeader);
+            if (usernameSecret != null) {
+                client_id = usernameSecret[0];
+                clientSecret = usernameSecret[1];
+            } else {
+                if (formData != null && !formData.containsKey(OAuth2Constants.CLIENT_ID)) {
+                    Response challengeResponse = Response.status(Response.Status.UNAUTHORIZED).header(HttpHeaders.WWW_AUTHENTICATE, "Basic realm=\"" + context.getRealm().getName() + "\"").build();
+                    context.challenge(challengeResponse);
+                    return;
+                }
+            }
+        }
+
+        if (formData != null) {
+            if (formData.containsKey(OAuth2Constants.CLIENT_ID)) {
+                client_id = formData.getFirst(OAuth2Constants.CLIENT_ID);
+            }
+            if (formData.containsKey(OAuth2Constants.CLIENT_SECRET)) {
+                clientSecret = formData.getFirst(OAuth2Constants.CLIENT_SECRET);
+            }
+        }
+
+        if (client_id == null) {
+            client_id = context.getSession().getAttribute("client_id", String.class);
+        }
+
+        if (client_id == null) {
+            Response challengeResponse = ClientAuthUtil.errorResponse(Response.Status.BAD_REQUEST.getStatusCode(), "invalid_client", "Missing client_id parameter");
+            context.challenge(challengeResponse);
+            return;
+        }
+
+        context.getEvent().client(client_id);
+
+        ClientModel client = context.getSession().clients().getClientByClientId(context.getRealm(), client_id);
+        if (client == null) {
+            context.failure(AuthenticationFlowError.CLIENT_NOT_FOUND, null);
+            return;
+        }
+
+        context.setClient(client);
+
+        if (!client.isEnabled()) {
+            context.failure(AuthenticationFlowError.CLIENT_DISABLED, null);
+            return;
+        }
+
+        if (client.isPublicClient()) {
+            context.success();
+            return;
+        }
```

Comment:
Is this authenticator run for all clients, or just our `uds-operator` client? Assuming it's only for our operator client, would it be better to fail + throw some error here because the client _shouldn't_ be public? Okay moving forward with this as-is though - I think this only plays a role if an admin manually changed the client to set it to public.

Reply:
No, this needs to stay as is. The Client Authenticators are called in a chain (chain of responsibility pattern) - one after another with the order defined in the Client Credentials Authorization Flow. If we block here, we essentially block other Client Authenticators from evaluating. 

In theory we could make it a bit more restrictive by calling `context.attempted()` but I decided to make it more permissive. This way, you can use it along with other Authenticators as "required" giving less chances for user misconfiguration.